javalib/src/main/scala/java/math/BigDecimal.scala

```diff
@@ -672,13 +672,27 @@ class BigDecimal() extends Number with Comparable[BigDecimal] {
 
   def multiply(multiplicand: BigDecimal): BigDecimal = {
     val newScale = this._scale.toLong + multiplicand._scale
+
+    @inline
+    def bigMultiply: BigDecimal = {
+      val unscaled = this.getUnscaledValue.multiply(multiplicand.getUnscaledValue)
+      new BigDecimal(unscaled, safeLongToInt(newScale))
+    }
+
+    import java.lang.{Long => JLong}
     if (this.isZero || multiplicand.isZero) {
       zeroScaledBy(newScale)
-    } else if (this._bitLength + multiplicand._bitLength < 64) {
-      valueOf(this._smallValue * multiplicand._smallValue, safeLongToInt(newScale))
+    } else if (this._bitLength + multiplicand._bitLength >= 64) {
+      bigMultiply
     } else {
-      val unscaled = this.getUnscaledValue.multiply(multiplicand.getUnscaledValue)
-      new BigDecimal(unscaled, safeLongToInt(newScale))
+      val x = this._smallValue
+      val y = multiplicand._smallValue
+      val z = x * y
+      if (JLong.signum(z) == JLong.signum(x) * JLong.signum(y)) {
```

Comment:
OK, but then my concern is that, as I said above, in general checking the signs is not enough to guarantee that no overflow occurs in a 2's complement multiplication (it is possible to multiply 2 negative numbers and receive a *negative* overflowed result). Why is it OK to do it like this in this instance? In other words, will any overflow happening when computing `z` be caught by this test, and why?